### PR TITLE
NAS-128996 / 24.10 / Improve filesystem apis (listdir and dosmode)

### PR DESCRIPTION
**Pull request by anodos325**
This commit adds the following improvements / fixes:

* filter_list is slightly tweaked to allow passing generators in addition to lists and tuples. This allows more memory-efficient filtering when we expect to exclude a large number of entries.

* business logic for generating middleware directory entries is moved to a helper class in src/middlewared/middlewared/utils/directory (DirectoryIterator) and tests are added.

* ZFS dosmode support is expanded to include coverage for full complement of ZFS file attributes and API renamed to get/set_zfs_attributes, and tests added for helper functions.

* helper function to determine whether a directory is empty is added to filesystem utils directory. Places where os.listdir was used directly and indirectly are replaced with `directory_is_empty` where it was deemed that directory may contain large number of files.

### CODE REVIEW

File: src/middlewared/middlewared/plugins/filesystem.py [hunk 1]
```diff
@@ -267,6 +263,8 @@ def listdir(self, path, filters, options):
           is_ctldir(bool): path is within special .zfs directory
```
Comment (bmeagherix) [inline src/middlewared/middlewared/plugins/filesystem.py:263]: Add comment line for `mount_id` too?

File: src/middlewared/middlewared/utils/__init__.py [hunk 2]
```diff
@@ -387,11 +387,11 @@ def filterop(self, i, f, source_getter):
 
         return False
 
-    def getter_fn(self, _list):
-        if not _list:
+    def getter_fn(self, entry):
+        if not entry:
```
Comment (bmeagherix) [inline src/middlewared/middlewared/utils/__init__.py:391]: Some docs / context would be helpful to facilitate review.

File: src/middlewared/middlewared/utils/filesystem/utils.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    is_in_ctldir = False
+    while path.as_posix() != '/':
+        if not path.name == '.zfs':
+            path = path.parent
+            continue
+
+        if path.stat().st_ino == ZFSCTL.INO_ROOT:
+            is_in_ctldir = True
+            break
+
```
[the remark this answers is not in the document]
Comment (anodos325) [reply src/middlewared/middlewared/utils/filesystem/utils.py:26]: No, that was a mistake on my part.

File: src/middlewared/middlewared/utils/filesystem/directory.py [hunk 4]
```diff
@@ -51,58 +52,144 @@ class DirectoryRequestMask(enum.IntFlag):
...
+        self.__dir_fd = os.open(path, os.O_DIRECTORY, dir_fd=dir_fd)
+
+    def __del__(self):
+        self.close()
+
+    def __repr__(self):
+        return f"<DirectoryFd path='{self.__path}' fileno={self.__fileno}>"
+
+    def close(self):
+        os.close(self.__dir_fd)
```
[the remark this answers is not in the document]
Comment (anodos325) [reply src/middlewared/middlewared/utils/filesystem/directory.py:75]: Derp, yes.

**Review by bmeagherix — APPROVED** [review]

**Pull request merged**